code/FOV QC/FOV QC utils.R

```diff
@@ -0,0 +1,297 @@
+    }
+    par(mar = c(0,0,2,0))
+    plot(res$xy, cex = 0.2, asp = 1, pch = 16,
+         col = colorRampPalette(c("darkblue", "blue", "grey80", "red", "darkred"))(101)[
+           pmax(pmin(51 + res$resid[match(res$gridinfo$gridid, rownames(res$resid)), i] * 50, 101), 1)], 
+         main = paste0(colnames(res$resid)[i], ": log2(fold-change)\nfrom comparable regions elsewhere"))
+    for (f in unique(res$fov)) {
+      inds <- res$fov == f
+      rect(min(xy[inds, 1]), min(xy[inds, 2]), max(xy[inds, 1]), max(xy[inds, 2]))
+    }
+    legend("right", pch = 16,
+           col = rev(c("darkblue", "blue", "grey80", "red", "darkred")),
+           legend = rev(c("< -1", -0.5, 0, 0.5, "> 1")))
+    if (!is.null(outdir)) {
+      dev.off()
+    }
+  })
+}
+
+
+#' Heatmap of estimated bit bias across FOVs
+#' 
+#' @param res Results object created by runFOVQC
+#' @return Draws a heatmap
+#' @export
+FOVEffectsHeatmap <- function(res) {
+  pheatmap::pheatmap(res$fovstats$bias * (res$fovstats$flag),
+                     col = colorRampPalette(c("darkblue", "blue", "white", "red", "darkred"))(100),
+                     breaks = seq(-2,2,length.out = 101),
+                     main = "FOV bias: log2(fold-change) from comparable regions in other FOVs")
+}
+
+
+
+
+#' Get nearest neighbors, sampling diffusely across other FOVs. 
+#' 
+#' @param x Data matrix, obs in rows, variables in columns
+#' @param fov vector of FOV IDs, aligned to rows of x
+#' @param n_neighbors How many neighbors to record
+#' @return A matrix of dimension nrow(obs) * n_neighbors, giving the row IDs for each obs's selected neighbors.
+getNearestNeighborsByFOV <- function(x, fov, n_neighbors = 10) {
+  # get nearest:
+  topneighbors <- FNN::get.knnx(data = x, 
+                                query = x, 
+                                k = min(n_neighbors * 50, nrow(x)))$nn.index
+  # reduce neighbors: none from own FOV, max of max_per_fov from each other fov:
+  reducedneighbors <- t(sapply(1:nrow(topneighbors), function(i) {
+    tempneighbors <- topneighbors[i, ]
+    # only accept 1 per FOV
+    tempneighbors[duplicated(fov[topneighbors[i, ]])] <- NA
+    # none from own FOV:
+    tempneighbors[fov[topneighbors[i, ]] == fov[i]] <- NA
+    return(tempneighbors[!is.na(tempneighbors)][1:n_neighbors])
+  }))
+  
+  return(reducedneighbors)
+}
+
+
+#' Get gridded expression within FOVs
+#' 
+#' @param mat Expression matrix, linear-scale
+#' @param xy 2-column matrix of xy locations
+#' @param fov Vector of FOV IDs
+#' @param squares_per_fov Number of squares to break fov into
+#' @param min_cells_per_square Fewest cells to keep a square
+#' @return A list: a matrix of expression in grid squares, a vector giving the grid square assignment of each cell.   
+makeGrid <- function(mat, xy, fov, squares_per_fov = 49, min_cells_per_square = 25) {
+  
+  # make grids:
+  ncuts <- floor(sqrt(squares_per_fov))
+  grid <- rep(NA, nrow(mat))
```

Comment:
Can get number of cells from `xy` and then don't need to provide expression matrix to this function

Reply:
good catch. fixed.